numba/tests/test_comprehension.py

```diff
@@ -248,14 +222,255 @@ def list24(x):
             bits = 64
         else:
             bits = 32
-        
+
         if utils.PYVERSION < (3, 0):
             with self.assertRaises(TypingError) as raises:
                 cfunc = jit(nopython=True)(list22)
                 cfunc(var)
             msg = "cannot unify reflected list(int%d) and int%d" % (bits, bits)
             self.assertIn(msg, str(raises.exception))
 
+class TestArrayComprehension(unittest.TestCase):
+
+    @tag('important')
+    def test_comp_with_array_1(self):
+        def comp_with_array_1(n):
+            m = n * 2
+            l = np.array([i + m for i in range(n)])
+            return l
+
+        cfunc = jit(nopython=True)(comp_with_array_1)
+        np.testing.assert_array_equal(comp_with_array_1(5), cfunc(5))
+        self.assertNotIn('allocate list', cfunc.inspect_llvm(cfunc.signatures[0]))
+
+        # test parallel array comprehension
+        cfunc = jit(nopython=True, parallel=True)(comp_with_array_1)
+        np.testing.assert_array_equal(comp_with_array_1(5), cfunc(5))
+        self.assertNotIn('allocate list', cfunc.inspect_llvm(cfunc.signatures[0]))
+        self.assertIn('@do_scheduling', cfunc.inspect_llvm(cfunc.signatures[0]))
+
+
+    @tag('important')
+    def test_comp_with_array_2(self):
+        def comp_with_array_2(n, threshold):
+            A = np.arange(-n, n)
+            return np.array([ x * x if x < threshold else x * 2 for x in A ])
+
+        cfunc = jit(nopython=True)(comp_with_array_2)
+        np.testing.assert_array_equal(comp_with_array_2(5, 0), cfunc(5, 0))
+        self.assertNotIn('allocate list', cfunc.inspect_llvm(cfunc.signatures[0]))
+
+    @tag('important')
+    def test_comp_with_array_noinline(self):
+        def comp_with_array_noinline(n):
+            m = n * 2
+            l = np.array([i + m for i in range(n)])
+            return np.sum(l)
+        import numba.inline_closurecall as ic
+        ic.enable_inline_arraycall = False
```

Comment:
Should this have some sort of fail management? If an exception is raised in the following `ic.enabled_inline_arraycall = False` persists?